django/core/handlers/asgi.py

```diff
@@ -24,6 +23,89 @@
 logger = logging.getLogger("django.request")
 
 
+class ASGIStream:
+    """
+    This class provides a minimalistic IO-like wrapper for an ASGI request
+    provided by a queue which is given as an asynchronous future
+    and its async event loop.
+    """
+
+    def __init__(self, receive):
+        self.buffer = bytearray()
+        self.receive = receive
+        self._has_more = True
+
+    @property
+    def has_more(self):
+        return bool(self._has_more or self.buffer)
+
+    def _receive_more_data(self):
+        if not self._has_more:
+            return b""
+
+        message = async_to_sync(self.receive)()
+        if message["type"] == "http.disconnect":
+            # Early client disconnect.
+            self._has_more = False  # safeguard against trying to call receive again
+            raise RequestAborted()
```

Comment:
This isn't triggered until the input queue is read at least once.

Reply:
As this is the intended behavior, maybe a better approach would be to test `ASGIStream` separately for this test case as a model test instead of the current integration test?

Reply:
Yes, happy to look at a refactor. 

But that integration test **should** pass no? (I **do** expect a timeout after a disconnect...?)

Reply:
We're not currently handling this correctly I think. Follow up at https://code.djangoproject.com/ticket/33738